pkg/virt-handler/device-manager/device_controller.go

```diff
@@ -87,23 +116,134 @@ func (c *DeviceController) startDevicePlugin(dev GenericDevice, stop chan struct
+					selector := removeSelectorSpaces(supportedMdev.Selector)
+					supportedMdevsMap[selector] = supportedMdev.ResourceName
+				}
+			}
+
+			hostMdevs := discoverPermittedHostMediatedDevices(supportedMdevsMap)
+			for mdevTypeName, mdevUUIDs := range hostMdevs {
+				mdevResourceName := supportedMdevsMap[mdevTypeName]
+				// add a device plugin only for new devices
+				if _, isRunning := c.devicePlugins[mdevResourceName]; !isRunning {
+					devicePluginsToRun[mdevResourceName] = ControlledDevice{
+						devicePlugin: NewMediatedDevicePlugin(mdevUUIDs, mdevResourceName),
+						stopChan:     make(chan struct{}),
+					}
+				} else {
+					delete(devicePluginsToStop, mdevResourceName)
+				}
+			}
+		}
+	}
+	return devicePluginsToRun, devicePluginsToStop
+}
+
+func removeSelectorSpaces(selectorName string) string {
+	// The name usually contain spaces which should be replaced with _
+	// Such as GRID T4-1Q
+	typeNameStr := strings.Replace(string(selectorName), " ", "_", -1)
+	typeNameStr = strings.TrimSpace(typeNameStr)
+	return typeNameStr
+
+}
+func (c *DeviceController) refreshPermittedDevices() {
 	logger := log.DefaultLogger()
-	logger.Info("Starting device plugin controller")
+	debugDevAdded := []string{}
+	debugDevRemoved := []string{}
+
+	// This function can be called multiple times in parallel, either because of multiple
+	//   informer callbacks for the same event, or because the configmap was quickly updated
+	//   multiple times in a row. To avoid starting/stopping device plugins multiple times,
+	//   we need to protect c.devicePlugins, which we read from in
+	//   c.updatePermittedHostDevicePlugins() and write to below.
+	c.devicePluginsMutex.Lock()
+
+	enabledDevicePlugins, disabledDevicePlugins := c.updatePermittedHostDevicePlugins()
+
+	// start device plugin for newly permitted devices
+	for resourceName, dev := range enabledDevicePlugins {
+		go c.startDevicePlugin(dev)
+		c.devicePlugins[resourceName] = dev
+		debugDevAdded = append(debugDevAdded, resourceName)
+	}
+	// remove device plugin for now forbidden devices
+	for resourceName, dev := range disabledDevicePlugins {
+		close(dev.stopChan)
+		delete(c.devicePlugins, resourceName)
+		debugDevRemoved = append(debugDevRemoved, resourceName)
+	}
 
+	c.devicePluginsMutex.Unlock()
+
+	logger.Info("refreshed device plugins for permitted/forbidden host devices")
+	logger.Infof("enabled device-pluings for: %v", debugDevAdded)
+	logger.Infof("disabled device-pluings for: %v", debugDevRemoved)
+}
+
+func (c *DeviceController) Run(stop chan struct{}) error {
+	logger := log.DefaultLogger()
+	// start the permanent DevicePlugins
 	for _, dev := range c.devicePlugins {
-		go c.startDevicePlugin(dev, stop)
+		go c.startDevicePlugin(dev)
 	}
+	c.virtConfig.SetConfigModifiedCallback(c.refreshPermittedDevices)
+	c.refreshPermittedDevices()
 
+	// keep running until stop
 	<-stop
 
+	// stop all device plugins
+	for _, dev := range c.devicePlugins {
+		dev.stopChan <- struct{}{}
+	}
```

Comment:
this makes me slightly nervous since i see we do a "close(dev.stopChan)" in another function. Do we need the ```c.devicePluginsMutex.Lock()``` here? From what I gather as long as a dev is in that slice, the stopChan isn't closed, but that's only guaranteed with using the mutex? 

Maybe this is safe after the `<-stop` though. just want to double check

Reply:
It's a good point. I just didn't expect to get a refresh when we are shutting down. 
I guess it's safer to use a lock here.